usr/opt/OSC200/usr/share/man/man4/fstab.4
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: fstab.4,v $ $Revision: 3.3.6.5 $ (DEC) $Date: 1993/11/23 20:12:06 $
.\" @(#)$RCSfile: fstab.4,v $ $Revision: 3.3.6.5 $ (DEC) $Date: 1993/11/23 20:12:06 $
.\" 
...\"
...\"  THIS OSF SOURCE FILE HAS BEEN MODIFIED BY DLIBRARY IN OSP
...\"  If you need information about the changes, send mail
...\"  to dlibrary@ravine at Digital Equipment Corp.
...\"
...\"
...\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
...\" ALL RIGHTS RESERVED
...\"
...\"
...\" Copyright (c) 1990, OPEN SOFTWARE FOUNDATION, INC.
...\" ALL RIGHTS RESERVED
...\"
...\" Copyright (c) 1980, 1989 The Regents of the University of California.
...\" All rights reserved.
...\"
...\" Redistribution and use in source and binary forms are permitted provided
...\" that: (1) source distributions retain this entire copyright notice and
...\" comment, and (2) distributions including binaries display the following
...\" acknowledgement:  ``This product includes software developed by the
...\" University of California, Berkeley and its contributors'' in the
...\" documentation or other materials provided with the distribution and in
...\" all advertising materials mentioning features or use of this software.
...\" Neither the name of the University nor the names of its contributors may
...\" be used to endorse or promote products derived from this software without
...\" specific prior written permission.
...\" THIS SOFTWARE IS PROVIDED \*LAS IS\*O AND WITHOUT ANY EXPRESS OR IMPLIED
...\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
...\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
...\"
...\" @(#)fstab.5	6.4 (Berkeley) 6/23/90
...\"
.TH fstab 4
.SH NAME
.PP
\*Lfstab\*O \- Static information about file systems and swap partitions
.SH SYNOPSIS
.PP
.sS
\*L/etc/fstab\*O
.sE
.SH DESCRIPTION
.PP
.iX "fstab file"
The
\*L/etc/fstab\*O
file contains descriptive information about the known file systems and the
disk partitions used for swapping (and dumping).
The \*L/etc/fstab\*O file
is read by various programs.  When you install the operating system,
the local (UFS) file systems and the partitions for the
swap areas are included
in the \*L/etc/fstab\*O file.
Each file system or swap partition is described on a separate line;
fields on each line are separated by tabs or spaces.
When you boot the system, only the file systems specified in the
\*L/etc/fstab\*O file are mounted.
.PP
The following is an example of an \*L/etc/fstab\*O file:
.oS
/dev/rz2a       /       ufs rw 1 1
/dev/rz0g       /usr    ufs rw 1 2
/dev/rz2b       swap1   ufs sw 0 2
/dev/rz0b       swap2   ufs sw 0 2
/dev/rz2g       /var    ufs rw 1 2
/dev/rz3c       /usr/users ufs rw 1 2
/usr/share/man@rabbit  /usr/share/man  nfs ro,bg 0 0
usr_dmn#user1   /usr/user1  advfs rw 0 0
.oE
.PP
The order of the lines in the \*L/etc/fstab\*O file
is important because the
\*Lfsck,\*O \*Lmount,\*O and \*Lumount\*O commands read the file
sequentially from top to bottom.
.PP
The syntax of a line in the \*L/etc/fstab\*O file is as follows.  Note
that lines beginning with a hash (\*L#\*O) sign are ignored.  Blank lines are
also ignored.
.PP
\*Vfile_spec\*O    \*Vmnt_point\*O    \*Vfs_type\*O  \*Vmnt_options\*O  \*Vbackup\*O  \*Vfsck\*O
.PP
The first field, (\*Vfile_spec\*O), describes the block special device,
the remote file system directory, or the AdvFS fileset to be mounted.
For UFS file systems, the special file name is the block special file name,
and not the character special file name.
If a program needs the character special file name,
the program must create it by appending the letter \*Lr\*O after the
last \*L/\*O (slash) in the special file name (for example, \*L/dev/rrz0g\*O).
For mfs file systems, \*Vfile_spec\*O can also specify the size in 512-byte
sectors, using the following syntax:
.oS
-s\*Vsize\*O
.oE
.PP
See \*Lmfs\*O(8) for more information.
.PP
The second field, (\*Vmnt_point\*O), specifies either the mount point for the
file system or remote directory or \*Lswap1\*O or \*Lswap2\*O for the primary
or secondary swap partition, respectively.
.\" Begin Digital-specific information
.\" I don't think this is true.  See above sentence.
.\" For swap partitions, this field should be specified as \*Lnone\*O.
.\" End Digital-specific information
.PP
The third field, (\*Vfs_type\*O), specifies the type of file system.
Specify \*Lufs\*O for swap partitions.
The system currently supports the following file systems:
.VL
.\" Begin Digital-specific information
.LI "\*Lcdfs\*O"
Specifies an ISO 9660 or High Sierra Formatted (CD\-ROM) file system.
.\" End Digital-specific information
.LI "\*Lnfs\*O"
Specifies a Network File System.
.\" Begin Digital-specific information
.LI "\*Lprocfs\*O"
Specifies a \*L/proc\*O file system, which allows you to access and manipulate
running processes as if they were files.  The \*L/proc\*O file system is
used for debugging purposes.  You must specify 0 (zero) in the
\*Vfreq\*O and \*Vorder\*O fields because the \*L/proc\*O
file system should not be backed up or checked.
.\" End Digital-specific information
.\" Begin Digital-specific information
.\" This is not true.  See above and below.
.\" .LI "\*Lswap\*O"
.\" Specifies a disk partition to be used for swapping.
.\" End Digital-specific information
.\" Begin Digital-specific information
.\" System V file systems not supported for Silver
.\" .LI "\*Ls5fs\*O"
.\" Specifies a System V file system.
.\" End Digital-specific information
.LI "\*Lufs\*O"
Specifies a local UNIX file system or a swap partition.
.\" Begin Digital-specific information
.\" Added mfs information due to change in support.
.LI "\*Lmfs\*O"
Specifies the memory file system.
.LI "\*Ladvfs\*O"
Specifies a local Advanced File System.
.\" End Digital-specific information
.LE
.PP
The fourth field, (\*Vmnt_options\*O),
describes the mount options associated with the file system or
partition.
It is formatted as a comma separated list of options and
contains at least the type of mount (see the \*Vfs_type\*O field)
plus any additional options appropriate to the file system type or the
partition use.  You can specify the following mount options:
.VL
.LI "\*Lro\*O"
Specifies that the file system is mounted with read-only access.
.\" Begin Digital-specific information CORRECTION
.\" Adding rq per QAR 9482.  This mount option is in fstab.h.
.LI "\*Lrq\*O"
Specifies that the file system is mounted with read-write access and quotas.
.\" End Digital-specific information CORRECTION
.LI "\*Lrw\*O"
Specifies that the file system is mounted with read-write access.
.LI "\*Luserquota\*O and \*Lgroupquota\*O"
Specifies the location of the quota files.  When the \*Lquota\*O
commands (for example, \*Lquotacheck\*O and \*Lquotaon\*O for UFS or
\*Lvquotacheck\*O and \*Lvquotaon\*O for AdvFS) are run, they first
check the mount points of the quota files specified by this option.
By default, user and group quotas for a file system are contained in
the \*Lquota.user\*O and \*Lquota.group\*O files, which are located in the
directory specified by the mount point.  For example, the quotas for the file
system on which \*L/usr\*O is mounted are located in the \*L/usr\*O
directory.  You also can specify another file name and location.
For example:
.sp
\*Luserquota=/var/quotas/tmp.user\*O
.sp
Note that quota options apply only to UFS and AdvFS file systems.
.LI "\*Lsw\*O"
Specifies that the partition is used as swap space.  If you specify the
\*Lsw\*O
mount option, you can also specify the following options that apply to
partitions used as swap space:
.RS
.VL
.LI "\*Lpri=\*Vn\*O"
Specifies the swap space priority.  The \*Vn\*O variable can be 0, 1, 2,
3, 4, with 0 being lowest priority, and 4 being highest priority.
.LI "\*Llowater=\*Vn\*O"
Specifies the low water mark. If the paging file grows larger
than the low water mark, and then shrinks below the low water mark,
the operating system will not make the file smaller than the low
water mark. If the low water mark is set to 0, then the paging file
will not shrink after paging space is freed.
The default value for the low water mark is 20Mbytes.
.LI "\*Lhiwater=\*Vn\*O"
Specifies the high water mark, which is the limit to which the
operating system will expand the paging file.
The default value for the high water mark is unlimited.
.LE
.RE
.LI "\*Lxx\*O"
Specifies that the file system entry should be ignored.
.LE
.\" Begin Digital-specific information
.\" The rq option is obsolete
.\" with the mount type rq (read/write with quotas)
.\" End Digital-specific information
.\" Begin Digital-specific information
.\" The rq option is obsolete
.\" If the \*Vfs_type\*O is specified as \*Lrq\*O the file system 
.\" is mounted as read/write with quotas. (\*Lrq\*O applies only to 
.\" the \*Lufs\*O file type.)
.\" End Digital-specific information
.PP
The fifth field, (\*Vbackup\*O), is used by the \*Ldump\*O
command to determine which file systems need to be backed up.
If the fifth field is not present, a value of zero is returned and
\*Ldump\*O
assumes that the file system does not need to be backed up.
.PP
The sixth field, (\*Vfsck\*O), is used by the
\*Lfsck\*O
command to determine the order in which file system checks are done
at reboot time.  For the root file system, specify \*L1\*O in the
\*Vfsck\*O field, and for other file systems, specify \*L2\*O
in the \*Vfsck\*O field.
File systems that are on the same disk are checked sequentially,
but file systems on different disks are checked at the
same time to utilize parallelism available in the hardware.
If the sixth field is not present or zero,
a value of zero is returned and the \*Lfsck\*O command
assumes that the file system does not need to be checked.
.PP
The following information is from the \*L/usr/include/fstab.h\*O
file:
.\" Begin Digital-specific information CORRECTION
.\" Restored, recoded, and moved the FSTAB_RQ line in response to QAR 7070.
.\" End Digital-specific information CORRECTION
.oS
#define	FSTAB_RW	"rw"	/* read/write device */
#define FSTAB_RQ        "rq"    /* read/write with quotas */
#define	FSTAB_RO	"ro"	/* read-only device */
#define	FSTAB_SW	"sw"	/* swap device */
#define	FSTAB_XX	"xx"	/* ignore totally */

struct fstab {
	char	*fs_spec;	/* block special device name */
	char	*fs_file;	/* file system path prefix */
	char	*fs_vfstype;	/* type of file system */
	char	*fs_mntops;	/* comma separated mount options */
	char	*fs_type;	/* rw, ro, sw, or xx */
	int	fs_freq;	/* dump frequency, in days */
	int	fs_passno;	/* pass number on parallel dump */
};
.oE
.PP
You can read records from the
\*L/etc/fstab\*O file
by using the \*Lgetfsent()\*O, \*Lgetfsspec()\*O, \*Lgetfstype()\*O,
and \*Lgetfsfile()\*O routines.
.SH RELATED INFORMATION
.PP
Files: \*L/usr/include/fstab.h\*O
.PP
Commands: \*Ladvfs\*O(4), \*Lgetfsent\*O(3), \*Lfsck\*O(8), \*Lmount\*O(8), \*Lumount\*O(8)
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/fstab.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char *rcsid = "@(#)$RCSfile: fstab.c,v $ $Revision: 4.2.9.3 $ (DEC) $Date: 1993/08/16 13:39:29 $";
#endif

/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
/*
 * Copyright (c) 1980, 1988 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 * fstab.c	5.8 (Berkeley) 8/21/89
 */


/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#if defined(_THREAD_SAFE)
#pragma weak endfsent_r = __endfsent_r
#pragma weak getfsent_r = __getfsent_r
#pragma weak getfsfile_r = __getfsfile_r
#pragma weak getfsspec_r = __getfsspec_r
#pragma weak setfsent_r = __setfsent_r
#endif
#if !defined(_THREAD_SAFE)
#pragma weak endfsent = __endfsent
#pragma weak getfsent = __getfsent
#pragma weak getfsfile = __getfsfile
#pragma weak getfsspec = __getfsspec
#pragma weak setfsent = __setfsent
#endif
#endif
#include <stdio.h>
#include <fcntl.h>
#include <fstab.h>

#include "ts_supp.h"

#define	MAXLINELENGTH	1024

#ifdef _THREAD_SAFE

extern char	*strtok_r();

#define	FS_FP			(*fs_fp)

#define	SETFSENT()		setfsent_r(fs_fp)
#define	FSTABSCAN(f, l)		fstabscan(f, l, len, FS_FP)
#define STRTOK(s1, s2)		strtok_r(s1, s2, &tokp)

#else

extern char	*strtok();

#define	FS_FP			fs_fp

#define	SETFSENT()		setfsent()
#define	FSTABSCAN(f, l)		fstabscan(f, l, MAXLINELENGTH, FS_FP)
#define STRTOK(s1, s2)		strtok(s1, s2)

static struct fstab	fs_fsent;
static FILE		*fs_fp;
static char		line[MAXLINELENGTH];

#endif	/* _THREAD_SAFE */


static int
fstabscan(struct fstab *fsent, char *line, int len, FILE *fp)
{
	register char *cp;
	int typexx;
	char subline[MAXLINELENGTH];
#ifdef _THREAD_SAFE
	char *tokp;
#endif	/* _THREAD_SAFE */
	char *fgets();

	for (;;) {
		/*
		 * Skip lines that start with hash sign
		 */
		do{
                        cp = fgets(line, len, fp);
                        if (cp == NULL)
                                return(0);
                } while (*cp == '#');

		/* original osf code */
/*
		if (!(cp = fgets(line, len, fp)))
			return (0);
*/
		fsent->fs_spec = STRTOK(cp, " \t\n");
		fsent->fs_file = STRTOK((char *)NULL, " \t\n");
		fsent->fs_vfstype = STRTOK((char *)NULL, " \t\n");
		fsent->fs_mntops = STRTOK((char *)NULL, " \t\n");
		if (fsent->fs_mntops == NULL)
			continue;
		fsent->fs_freq = 0;
		fsent->fs_passno = 0;
		if ((cp = STRTOK((char *)NULL, " \t\n")) != NULL) {
			fsent->fs_freq = atoi(cp);
			if ((cp = STRTOK((char *)NULL, " \t\n")) != NULL)
				fsent->fs_passno = atoi(cp);
		}
#ifdef _THREAD_SAFE
		/* paranoia - check for a very long list of mntops */
		if (strlen(fsent->fs_mntops) > MAXLINELENGTH)
			continue;
#endif	/* _THREAD_SAFE */
		strcpy(subline, fsent->fs_mntops);
		for (typexx = 0, cp = STRTOK(subline, ","); cp;
		     cp = STRTOK((char *)NULL, ",")) {
			if (strlen(cp) != 2)
				continue;
			if (!strcmp(cp, FSTAB_RW)) {
				fsent->fs_type = FSTAB_RW;
				break;
			}
			if (!strcmp(cp, FSTAB_RQ)) {
				fsent->fs_type = FSTAB_RQ;
				break;
			}
			if (!strcmp(cp, FSTAB_RO)) {
				fsent->fs_type = FSTAB_RO;
				break;
			}
			if (!strcmp(cp, FSTAB_SW)) {
				fsent->fs_type = FSTAB_SW;
				break;
			}
			if (!strcmp(cp, FSTAB_XX)) {
				fsent->fs_type = FSTAB_XX;
				typexx++;
				break;
			}
		}
		if (typexx)
			continue;
		if (cp != NULL)
			return (1);
	}
	/* NOTREACHED */
}


#ifdef _THREAD_SAFE
int
getfsent_r(struct fstab *fsent, char *line, int len, FILE **fs_fp)
{
#else
struct fstab *
getfsent(void)
{
	register struct fstab	*fsent = &fs_fsent;
#endif	/* _THREAD_SAFE */

	TS_EINVAL((fsent == 0 || line == 0 || len <= 0 || fs_fp == 0));

	if (!FS_FP && SETFSENT() != TS_SUCCESS || !FSTABSCAN(fsent, line))
		return (TS_NOTFOUND);
	return (TS_FOUND(fsent));
}


#ifdef _THREAD_SAFE
int
getfsspec_r(const char *name, struct fstab *fsent,
	    char *line, int len, FILE **fs_fp)
{
#else
struct fstab *
getfsspec(register const char *name)
{
	register struct fstab	*fsent = &fs_fsent;
#endif	/* _THREAD_SAFE */

	TS_EINVAL((fsent == 0 || line == 0 || len <= 0 || fs_fp == 0));

	if (SETFSENT() == TS_SUCCESS)
		while (FSTABSCAN(fsent, line))
			if (!strcmp(fsent->fs_spec, name))
				return (TS_FOUND(fsent));
	return (TS_NOTFOUND);
}


#ifdef _THREAD_SAFE
int
getfsfile_r(const char *name, struct fstab *fsent,
	    char *line, int len, FILE **fs_fp)
{
#else
struct fstab *
getfsfile(register const char *name)
{
	register struct fstab	*fsent = &fs_fsent;
#endif	/* _THREAD_SAFE */

	TS_EINVAL((fsent == 0 || line == 0 || len <= 0 || fs_fp == 0));

	if (SETFSENT() == TS_SUCCESS)
		while (FSTABSCAN(fsent, line))
			if (!strcmp(fsent->fs_file, name))
				return (TS_FOUND(fsent));
	return (TS_NOTFOUND);
}


#ifdef _THREAD_SAFE
int
setfsent_r(FILE **fs_fp)
#else
int
setfsent(void)
#endif	/* _THREAD_SAFE */
{
	int	flags;

	TS_EINVAL((fs_fp == 0));

	if (FS_FP) {
		rewind(FS_FP);
		return (TS_SUCCESS);
	}

	if ((FS_FP = fopen(_PATH_FSTAB, "r")) != NULL) {
		flags = fcntl(fileno(FS_FP), F_GETFD, 0);
		flags |= FD_CLOEXEC;
		if (fcntl(fileno(FS_FP), F_SETFD, flags) == 0)
			return (TS_SUCCESS);
		(void)fclose(FS_FP);
	}
	return (TS_FAILURE);
}


#ifdef _THREAD_SAFE
void
endfsent_r(FILE **fs_fp)
#else
void
endfsent(void)
#endif	/* _THREAD_SAFE */
{
#ifdef _THREAD_SAFE
	if (fs_fp == 0) return;
#endif	/* _THREAD_SAFE */

	if (FS_FP) {
		(void)fclose(FS_FP);
		FS_FP = NULL;
	}
}
